File: 0605-can-place-flowers/0605-can-place-flowers.py
class Solution:
    def canPlaceFlowers(self, flowerbed: List[int], n: int) -> bool:
        memo = {}
        def dp(i, prev):
            if (i, prev) in memo:
                return memo[(i, prev)]

            if i >= len(flowerbed):
                return 0
            
            next = flowerbed[i + 1] if i + 1 < len(flowerbed) else 0
            current = flowerbed[i]

            ans = memo[(i, prev)] = max(1 + dp(i + 2, next), dp(i + 1, 0)) if prev + next + current == 0 else dp(i + 1, current)

            return ans
        return dp(0, 0) >= n

File: 0605-can-place-flowers/0605-can-place-flowers.cpp
class Solution {
public:
    bool canPlaceFlowers(vector<int>& flowerbed, int n) {
        int count = 0;
        bool skip = false;
        int ans = 0;

        for (auto const& x : flowerbed) {
            if (skip) {
                skip = false;
                continue;
            }

            if (x == 1) {
                ans += count / 2;
                count = 0;
                skip = true;
            } else {
                count += 1;
            }
        }

        return ans + (count + 1) / 2 >= n;   
    }
};